core/chain/src/lib.rs

```diff
@@ -96,6 +123,27 @@ where
         self.update_best_block(block);
     }
 
+    /// Inserts a verified header.
+    pub fn insert_header(&self, header: B::SignedHeader) {
+        let block_hash = header.block_hash();
+        if self.is_known_header(&block_hash) {
+            return;
+        }
+
+        if self
+            .storage
+            .write()
+            .expect(POISONED_LOCK_ERR)
+            .blockchain_storage_mut()
+            .header(&header.parent_hash())
+            .unwrap()
+            .is_none()
+        {
+            return;
```

Comment:
Should we store orphaned header somewhere?

Reply:
This suppose to be "valid" header. E.g. have parents. 
We already store orphaned in the syncing/importing code.